aeternity/epoch
# Pt 158520766 contract poi tests

Corresponding spec PR: https://github.com/aeternity/protocol/pull/128

TODO:
* [ ] Re-target on master once PR #1275 merged
* [x] Stricter prefix iterator testing
* [ ] ... see https://www.pivotaltracker.com/story/show/158520766
  * [x] Test broken POI
    * Tested by removal of one of the contract-related nodes in the PoI
  * [x] Test a POI where some of the contract state is missing
    * See test case "PoI for one contract without store that becomes with store"
  * [x] Test a POI where there is an extra contract state key/value
    * See test case "PoI for one contract with store that becomes without store"
  * [ ] Test corner cases with empty keys etc.
    * Done though I plan to address case "verification of PoI with a contract with an empty (hence invalid) key" in a follow-up PR.
* [x] Clarify whether empty key can be in contract store

Maybe TODO:
* [x] Stricter assertion on contract state set e.g. both key and value are binaries
* [ ] Clarify whether empty value can be in contract store
  * Tracked as https://www.pivotaltracker.com/story/show/158630078
* [ ] Clarify how [`lookup` third argument](https://github.com/aeternity/epoch/blob/v0.16.0/apps/aeutils/src/aeu_mp_trees.erl#L177) works in PoI verification so to understand whether the value can be removed from the verify function.
